0023-communication-protocol.adoc

```diff
@@ -0,0 +1,383 @@
+
+We can define very simple protocols to allow for this information to be exchanged.
+These are the protocols we need:
+
+==== Protocols for sharing identities
+
+Protocol identifier: `/comit/swap/identity/bitcoin/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79", // <1>
+  "pubkey": "03c9c6a4cd77d03ebe552e85c2a84a5065263824cb006629659821a0bbe496538e" // <2>
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. A compressed bitcoin public key (33 bytes)
+
+Protocol identifier: `/comit/swap/identity/ethereum/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79", // <1>
+  "address": "0xd3935B8cCA773146Cda591eF468Ce141BD6d57f4" // <2>
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. An Ethereum address
+
+Protocol identifier: `/comit/swap/identity/lightning/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79", // <1>
+  "pubkey": "0236055ab508ec9495c07a3968f7ee9a1b5983f70e24d362f677d2619419601052" // <2>
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. A Lightning node public key (taken from `lncli getinfo`)
+
+==== Why do we not need more protocols?
+
+Given the context of a swap, all of these protocols carry unambiguous information.
+Consider the following example:
+
+Assume `Bitcoin` is the alpha-ledger and `Ethereum` is the beta-ledger.
+In said constellation, Alice's refund identity is a Bitcoin public key and here redeem identity is an Ethereum address.
+
+If Alice would send a message using the `/comit/swap/identity/bitcoin/1.0.0` protocol, it is unambiguous for Bob that Alice is sending her refund identity.
+
+More generally speaking:
+Given knowledge of the alpha and beta ledger and the role, a message can be unambiguously identified as the redeem or refund identity of an actor depending on whether the message is incoming or outgoing.
+
+This allows us to slim down the protocols to the bare minimum of sending over an identity.
+
+=== Finalizing a swap
+
+One of the downsides of splitting the communication for a single swap up into several protocols is the fact that we don't have a natural "end" of the communication.
+If all of the communication were to take place within a single protocol, closing the substream could act as "I am done communicating".
+
+Following a multiple protocol approach, the answer to how we communicate this concern is another protocol.
+
+Introducing: `/comit/swap/finalize/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79"
+}
+----
+
+Both parties are expected to use this protocol after they have sent all the data they consider necessary for starting the execution of the swap.
+Once each party sent a finalize message AND received one, implementations should consider the swap finalized and start with the execution phase (i.e. watch for transactions, emit actions, etc)
```

Comment:
I assume that finalize should involve some finale verification? 
Let's say Alice sends `finalize` to Bob, he verifies if he has all information he needs and then returns `finalize` to Alice.

Reply:
No, finalize doesn't involve any information.
The verification was done using the swap_digest.

Regarding the identities, I think it is reasonable to assume that they didn't mess them up.

The reasoning for the finalize message is to make sure, both parties know that there are not going to be any more messages.

Reply:
Sorry, I wasn't clear: I meant that Bob verifies that he has all information *he needs from Alice* before answering `finalize` to Alice.
 
The protocol proposed works the other way around so far: Alice tells Bob what she thinks he would need. If there is however a problem or a mismatch (an error occurred, protocol update, etc), then it might be the case that Bob would need more information from Alice. In this case, he should nor answer `finalize` to Alice.

Reply:
Not quite:
Both parties tell the other parties what they think they will need.

The idea is that each party doesn't send `finalize` until:

- they sent all the data they intend to send
- they received all the data they need

If the finalize handshake doesn't happen within a certain time period, we should timeout and abort the swap (there seems to be a bug in the implementations then).

Reply:
Note that the new protocol doesn't have request-reply but just one-way messages. They also all happen on different substreams, meaning it all happens concurrently.